Bugfix/copy listener buffer

DonKingDingaLing opened the pull request:
Changed IsotpListener to return a copy of the buffer. 
Otherwise multi-threading applications may run into the problem of getting the buffer written to while the previous message is still being processed. 

Maybe should be backported to version 2.X.X aswell.

pschichtel commented:
Seems like CI is broken for PRs, so we can ignore that.

I wonder if this should be an option. I originally didn't design the listener for direct multi threading. You may have noticed that the EventLoop class backing the listener is hardcoded to a single IO thread for polling. The handlers are executed on that thread and so further processing of events (and thus new writes into that buffer) will be blocked until the handler is finished. So any downstream handler should, if needed for further processing, be able to safely copy the buffer itself before handing it over to another thread.

Can you elaborate on the setup (specifically when it comes to threading) in which you encountered this issue?

DonKingDingaLing commented:
Thanks for the quick response and sorry for my late response. 
I tried to provide you with a test for it but I struggle with JUnit in combination with threading today.
So I'll try explain it. 

When the event loop polls it acts in a seperate thread which makes perfect sense. 
Within this thread the polling mechanism calls the MessageHandler. 
In our case the MessageHandler just stores the polled message somewhere, so that another thread can pick it up later. 

The problem with this is that when we store the ByteBuffer somewhere for later use, we actually just store a reference to the ByteBuffer of the IsotpListeners ByteBuffer. 

Like you said, this can be avoided by just copying it. 
I think however that it avoids errors if that copying is being done in the library it self. 
I could imagine that other java developers could run into this problem aswell otherwise. 

So to conclude, it's not really a bug in that way that it breaks stuff activly, but it's somewhat breaking with java mindset and therefore could lead to issues.

pschichtel commented:
Yeah I see the point that the current behavior might lead to memory-unsafe usage, especially since the threading model of these helpers is not particularly well documented at the moment. On the other side adding the memory copy in this place will add overhead for all users including those that use the buffer safely.

So I think I'd prefer one of these approaches:

1. Just document the threading model and the way the buffer is handled and reused
2. Provide a MessageHandler proxy that copies the buffer before passing it on
3. Add an option to the listeners so that buffer copying can be turned on and off per listener instance

I think out of these approaches a combination of 1 and 2 would make the most sense. wdyt?

DonKingDingaLing commented:
I like the idea. I have updated the PR accordingly. I've introduced a override to the addChannel method. In this overload a flag can be set, which is stored per channel. Depending on that flag during the procession of an event can be decided whether or not to hand over a copy. Also I added a line on the doc comment of addChannel to emphasis that a reference will be handed over.

pschichtel commented:
what do you think about this:

https://github.com/pschichtel/JavaCAN/blob/master/epoll/src/main/java/tel/schich/javacan/util/CopyingMessageHandlerProxy.java

I've also added some documentation on the listener

DonKingDingaLing commented:
sorry forgot to answer you on friday. 

yeah I like it, better than mine adds something without changing existing stuff =)

pschichtel commented:
OK cool! So if you are happy with that than I'll close the PR. Thanks a lot for your contribution anyway!

DonKingDingaLing commented:
@pschichtel yeah a backport would be quite nice.

pschichtel commented:
Does that imply you are using the 2.x branch? If so, may I ask if there is a reason for that?

DonKingDingaLing commented:
In all honesty, we didn't want to compile the library for our architecture at this stage of our project, because 2.X already contains the fitting library for our architecture. That being said we plan on migrating to 3.x in the future.

pschichtel commented:
Ah ok, good to hear that the upgrade is planned. JavaCAN 3.x supports all the archnitectures out of the box that 2.x supports, quite a few more actually. 3.x got rid of the architecture detection since it turned out broken and unfixable in many cases, but the prebuilt architecture are all still there.

Anyway, I'll release new version for 3.x and 2.x in a few minutes

pschichtel commented:
there you go: https://github.com/pschichtel/JavaCAN/releases/tag/javacan-2.3.1

DonKingDingaLing commented:
Awesome thank you so much =)

DonKingDingaLing commented:
Sorry to be so annoying, but could you maybe release the 3.2.1 and 2.3.1 artifact to maven. Apperently the pipeline didn't push them.

pschichtel commented:
The pipeline does not publish release builds, I did that locally

pschichtel commented:
Seems you are right, the deployment of the 3.2.1 failed with a validation error. Sadly I'm not able to redo that deployment until end of April. 2.3.1 should be fine

pschichtel commented:
@DonKingDingaLing almost forgot about re-releasing 3.2.1, but I just got to it: https://github.com/pschichtel/JavaCAN/releases/tag/javacan-3.2.2

pschichtel commented:
@DonKingDingaLing just stumbled over this again: Latest JavaCAN 3.3.x has once again architecture auto detection, that might make porting from 2.x easy for you.
